mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,909 @@
+// exist. This case differs from the CutCase in that it can still coalesce
+// polytopes, as long as they are not sticking out of each other by too much
+// (i.e. by less than 2). As we are considering an integer setting, the convex
+// hull of the union of those two polytopes actually is the same thing as the
+// union of the two polytope.
+//      _____          _____
+//    _|___  |        /     |
+//   | |   | |   ==> |      |
+//   | |   | |       |      |
+//   |_|___|_|       |______|
+//
+// TODO: find better example
+bool protrusionCase(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    Info &info_a, const Info &info_b, unsigned i, unsigned j);
+
+// compute, whether a constraint of cut sticks out of bs by more than 2
+bool stickingOut(const SmallVector<ArrayRef<int64_t>, 8> &cut,
+                 const FlatAffineConstraints &bs);
+
+// adds a FlatAffineConstraints and removes the sets at i and j.
+void addCoalescedBasicSet(
+    SmallVectorImpl<FlatAffineConstraints> &basicSetVector, unsigned i,
+    unsigned j, const FlatAffineConstraints &bs);
+
+// compute the cut case and return whether it has worked.
+//
+// The cut case is the case, for which a polytope only has REDUNDANT and CUT
+// constraints. If all the facets of such cut constraints are contained within
+// the other polytope, the polytopes can be combined to a polytope only
+// limited by all the REDUNDANT constraints.
+// _________          _________
+// \ \  |  /          \       /
+//  \ \ | /    ==>     \     /
+//   \_\|/              \___/
+//
+//
+bool cutCase(SmallVector<FlatAffineConstraints, 4> &basicSetVector, unsigned i,
+             unsigned j, const Info &info_a, const Info &info_b);
+
+// compute adj_ineq pure Case and return whether it has worked.
+//
+// The adj_ineq pure case can be viewed as a single polytope originally (below:
+// on the right), that was cut into two parts by a strip with width 1.
+// This is computed by just using all the REDUNDANT constraints.
+//  ________ ____            ______________
+// |       //    |          |              |
+// |      //     |          |              |
+// |     //      |    ==>   |              |
+// |    //       |          |              |
+// |___//________|          |______________|
+//
+bool adjIneqPureCase(SmallVector<FlatAffineConstraints, 4> &basicSetVector,
+                     unsigned i, unsigned j, const Info &info_a,
+                     const Info &info_b);
+
+// compute the non-pure adj_ineq case and return whether it has worked.
+// Constraint t is the adj_ineq.
+//
+// In the non-pure adj_ineq case, one of the polytopes is like an extension of
+// the other one. This can be computed, by inverting the adj_ineq and checking
+// whether all constraints are stll valid for this new polytope.
+//   ____          ____
+//  |    \        |    \
+//  |     \   ==> |     \
+//  |_____|_      |______\
```

Comment:
The second polytope is not clearly visible in this diagram. Please write a line pointing out where it is.

Reply:
I added:
In the example below, Polytope A is the big one on the left, Polytope B is the small line at the right side. that gets included after coalescing.